bgp: withdraw route when svc has no endpoints

oblazek opened the pull request:
We have services with either externalTrafficPolicy or internalTrafficPolicy. We have discussed in the issue https://github.com/cilium/cilium/issues/40438 that according to the docs - externalTrafficPolicy applies to loadBalancerIP and/or externalIP and internalTrafficPolicy applies to clusterIP. There was also a comment that it doesn't matter where the client is located as e.g. loadBalancerIP can also be accessed from within the cluster (from e.g. pod) even though it's not that common. 
When endpoints are taken into account the value of the externalTP or internalTP  matters. In BGP world when either is set to `Local`, the IP (LBIP, externalIP or clusterIP) is announced only if the node hosts an endpoint. When eTP/iTP is set to the other value - `Cluster`, then all nodes will announce the IP even though they do not host any relevant pod/endpoint.

In the current codebase the code doesn't handle properly when service has zero endpoints and the eTP/iTP is set to `Cluster`. In this case only for `Local` the node will stop announcing the IP and the packet targetted to the service will be dropped [here](https://github.com/cilium/cilium/blob/main/bpf/bpf_sock.c#L290-L291). The problem is that this doesn't take eTP/iTP value into account and dropps the packet in both cases.

[We've agreed](https://github.com/cilium/cilium/issues/40438#issuecomment-3090582533) that for eTP/iTP == `Cluster` 
we will not drop the packet, but will forward it to the stack. Aditionally our cilium bgp subsystem will stop announcing the IP so that setups that use e.g. anycast can benefit from this (and traffic will be forwarded to e.g. another DC which [announces the same VIP](https://github.com/cilium/cilium/issues/40438#issuecomment-3053556535)).   

This PR handles this scenario.

TLDR:
* add a new flag which defaults to true, so that current behavior is not changed
* keep track of services when corresponding epDiffstore objects has been updated
* withdraw a route when service has 0 endpoint newly for eTP/iTP=Cluster
* socklb - do not drop packets when service has 0 endpoints and eTP/iTP=Cluster

Fixes: [#40438](https://github.com/cilium/cilium/issues/40438)

cc @rastislavs 

```release-note
bgp: optionally withdraw route when service has 0 endpoints to allow balancing to a different DC/cluster (anycast)
```

rastislavs commented on pkg/bgpv1/manager/reconciler/service.go:
```diff
@@ -24,24 +25,34 @@ import (
 	ciliumslices "github.com/cilium/cilium/pkg/slices"
 )
 
+type RoutesConfig struct {
```
This config seems to be defined at 2 places now. What about defining a common `pkg/bgpv1/config` and using it from there? In the future, we can move there other BGP-specific config options, such as `bgp-secrets-namespace`

oblazek replied:
sounds good

rastislavs commented on pkg/bgpv1/cell.go:
```diff
@@ -91,6 +91,10 @@ var Cell = cell.Module(
 	reconciler.ConfigReconcilers,
 
+	cell.Config(reconcilerv2.RoutesConfig{
+		BGPNoEndpointsRoutable: option.Config.BGPNoEndpointsRoutable,
```
Could we define something like `DefaultRoutesConfig` and provide that into the hive, like in this example: https://docs.cilium.io/en/latest/contributing/development/hive/#config ?

oblazek replied:
thanks, did that

asauber reviewed (approved):
Approving for config changes

rgo3 commented on pkg/datapath/linux/config/config.go:
```diff
@@ -341,6 +341,10 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 		}
+
+		if option.Config.BGPNoEndpointsRoutable {
+			cDefinesMap["NO_ENDPOINTS_ROUTABLE"] = "1"
```
Could we use the new Cilium config mechanism for this datapath config? You can take a look at https://docs.cilium.io/en/stable/contributing/development/datapath_config/ for reference. Take a look at e.g. `bpf/include/bpf/config/lxc.h` and create something analog to this for bpf_sock.c to use `DECLARE_CONFIG` for your new config.

oblazek replied:
totally! thanks I had no idea there is a new way

oblazek replied:
done :)

bimmlerd commented on pkg/option/config.go:
```diff
@@ -992,6 +992,9 @@ const (
 	EnableBGPControlPlaneStatusReport = "enable-bgp-control-plane-status-report"
 
+	// BGPNoEndpointsRoutable enables advertising of services with 0 endpoints
+	BGPNoEndpointsRoutable = "bgp-no-endpoints-routable"
```
nit: I don't have context on this change, but in general adding to `option/config` is no longer the way to go. Is there a way these can move into config cells instead?

oblazek replied:
yeah totally, let me see if I am able to pass the tests without this.. maybe I am missing something

oblazek replied:
but will have to figure out how do I pass that to the `pkg/datapath/config...`

oblazek replied:
done, ptal :)

bimmlerd replied:
Nice, thank you. If you mark the PR as ready for review I can give a LGTM - otherwise the approval won't count once you mark it ready, because GitHub 🤷‍♂️

oblazek committed:
bgp: withdraw a service when there are no endpoints

Signed-off-by: Ondrej Blazek <ondrej.blazek@firma.seznam.cz>

oblazek committed:
socketLB: include svcRouteConfig package

Previously it was not possible to access services with 0 endpoints via
socketLB and all packets were dropped with EHOSTUNREACH.

With these changes it is possible to access services with 0 endpoints
which can resolve in traffic being forwarded to another DC that can
serve the same service (e.g. one with the same loadbalancerIP).

Signed-off-by: Ondrej Blazek <ondrej.blazek@firma.seznam.cz>

oblazek committed:
docs: mention new enable-no-endpoint-routes flag

* update helm to include enableNoServiceEndpointsRoutable top level flag

Signed-off-by: Ondrej Blazek <ondrej.blazek@firma.seznam.cz>

oblazek committed:
test: add testfile for enable-no-endpoints-routable=false

Signed-off-by: Ondrej Blazek <ondrej.blazek@firma.seznam.cz>

oblazek committed:
bgp: include svcrouteconfig pkg in reconciler unit tests

Signed-off-by: Ondrej Blazek <ondrej.blazek@firma.seznam.cz>

oblazek committed:
pkg: add svcrouteconfig package

Add a new flag `--enable-no-endpoints-routable` to enable routing
to services with zero endpoints.
This can later be imported by e.g. socketLB and/or BGP to withdraw
routes when service has no active endpoint.

Signed-off-by: Ondrej Blazek <ondrej.blazek@firma.seznam.cz>

oblazek committed:
codeowners: add svcrouteconfig package

Signed-off-by: Ondrej Blazek <ondrej.blazek@firma.seznam.cz>

gandro commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -777,6 +777,10 @@ data:
 {{- end }}
 
+{{- if hasKey .Values "enableNoEndpointsRoutable" }}
+  enable-no-endpoints-routable: {{ .Values.enableNoEndpointsRoutable | quote }}
```
We should document this in the values.yaml.tmpl. Also consider not making this a top-level value, but add it as a bgp subvalue, e.g. `bgp.enableNoEndpointsRoutable`.

I also wonder if we should prefix the agent flag with `bgp`, since it isn't clear without context that `enable-no-endpoints-routable` refers to a BGP behavior.

oblazek replied:
I was hesitant about this since this also modifies socketLB behaviour not strictly BGP only, but it makes more sense to me as well

rastislavs replied:
I am torn whether this should be BGP specific config or not.. 

> this also modifies socketLB behaviour not strictly BGP only

Is there any use-case for enabling `enable-no-endpoints-routable` outside of the BGP one?

oblazek replied:
Yeah, might be not that common, but when bgpControlPlane is not enabled and the classic services of type LoadBalancer provided by external loadbalancer there still will be svc entry in a bpf map, so that packet can be decapsulated and a "socket" found. When accessed from within the cluster the code will end up in exactly the same place and it won't matter if BGP is enabled or not.

oblazek replied:
but no objection in adding the bgp prefix

rastislavs replied:
> When accessed from within the cluster the code will end up in exactly the same place and it won't matter if BGP is enabled or not

Right, because of this (potential use of external loadbalancer other than BGP) I would slightly prefer to keep the agent flag as it is (= not prefixed with BGP). I guess that's also the reason why you moved the config to a package outside of the bgp code tree (`pkg/svcrouteconfig/cell.go`).

gandro replied:
I'm also fine with keeping it top-level if it affects SockLB too. My only other suggestion would be to clarify that we're talking about services, e.g. `enableNoServiceEndpointsRoutable` because the term "endpoint" is very overloaded in Cilium (i.e. there are also the concept of a Cilium endpoints. There's a feature called endpoint routes, which I originally though this related to, which is why I suggested adding the BGP context).

But I don't feel strongly at all about this, just bikeshedding here, so feel free to ignore.

oblazek replied:
no you are right, it might sound like endpoint-routes which is really not related, let me add the `Service` => `enableNoServiceEndpointsRoutable`

gandro reviewed (commented):
Reviewed Helm changes

rastislavs reviewed (approved):
BGP changes LGTM!

rgo3 commented on pkg/datapath/config/gen.go:
```diff
@@ -13,3 +13,4 @@ package config
 //go:generate go run github.com/cilium/cilium/tools/dpgen -path ../../../bpf/bpf_overlay.o -embed Node -kind object -name BPFOverlay -out overlay_config.go
 //go:generate go run github.com/cilium/cilium/tools/dpgen -path ../../../bpf/bpf_network.o -embed Node -kind object -name BPFNetwork -out network_config.go
 //go:generate go run github.com/cilium/cilium/tools/dpgen -path ../../../bpf/bpf_wireguard.o -embed Node -kind object -name BPFWireguard -out wireguard_config.go
+//go:generate go run github.com/cilium/cilium/tools/dpgen -path ../../../bpf/bpf_sock.o -kind object -name BPFSocket -out sock_config.go
```
Lets keep the `Node` object embedded as with the other configuration structs (mainly for future use) and keep the naming in line with the C/.o bpf files:
```suggestion
//go:generate go run github.com/cilium/cilium/tools/dpgen -path ../../../bpf/bpf_sock.o -embed Node -kind object -name BPFSock -out sock_config.go
```

oblazek replied:
understood, I'm on it

oblazek replied:
done, but had to export/move `nodeConfig()` function to avoid import cycle loop :/

rgo3 reviewed (approved):
LGTM, assuming my open comment will be addressed. Thanks @oblazek!

gandro reviewed (approved):
Looks good for Helm!

marseel reviewed (approved):
Thanks!

oblazek committed:
datapath: embed Node into BPFSocket

Signed-off-by: Ondrej Blazek <ondrej.blazek@firma.seznam.cz>

joestringer reviewed (commented):
The changes I see look fine, but two minor points:

1. The release note in the PR description implies that this changes the default setting to withdraw routes, but the rest of the text implies there is no change to the default. Could you clarify that in the release note? When users read the release note, they may read that note and take away that there was a change to this behavior by default when there isn't.

2. I was requested review as a member of @cilium/docs-structure , but this new functionality doesn't appear to have any docs. Not sure if there's a good place to document this new behavior but did you consider how to explain this feature to other users?

joestringer reviewed (approved):
LGTM, thanks.

aditighag reviewed (changes requested):
Hi! Couple of clarifying questions - 
(1) >socketLB: include svcRouteConfig package
Can you add context to the commit/PR description why this change is only needed in socket-LB and not per packet-LB? We have `DROP_NO_SERVICE` drops in per-packet LB datapath as well. 
(2) The PR mentions that when there was no service endpoints, packets would be routed by BGP installed routes. This suggests that the config needs to be enabled only when BGP control plane is enabled, no?

oblazek committed:
datapath: add svcrouteConfig to LNC

* update lxc, nodeport logic around NO_SERVICE_BACKEND with new flag

Signed-off-by: Ondrej Blazek <ondrej.blazek@firma.seznam.cz>

oblazek committed:
bpf: define enable_no_service_endpoints_routable in tests

Signed-off-by: Ondrej Blazek <ondrej.blazek@firma.seznam.cz>

aditighag commented on bpf/bpf_lxc.c:
```diff
@@ -127,14 +127,27 @@ static __always_inline int __per_packet_lb_svc_xlate_4(void *ctx, struct iphdr *
-
-		if (IS_ERR(ret))
+				}
+			}
```
I think moving this inner nested logic to a helper could improve code readability and reuse. You can invoke the same helper down below in nodeport handling code path.

oblazek replied:
sounds good, thanks, will do

oblazek replied:
done

aditighag reviewed (commented):
The job checking every commit compiles is failing -- can you take a look?

aditighag commented on bpf/lib/lb.h:
```diff
@@ -2420,3 +2432,29 @@ __wsum icmp_wsum_accumulate(void *data_start, void *data_end, int sample_len)
 #endif /* SERVICE_NO_BACKEND_RESPONSE */
+
+static __always_inline
+int handle_nonroutable_endpoints_v4(struct lb4_service *svc)
+{
+	if ((lb4_svc_is_external(svc) &&
+	     (svc->flags & SVC_FLAG_EXT_LOCAL_SCOPE)) ||
+	   (!lb4_svc_is_external(svc) &&
+	    (svc->flags2 & SVC_FLAG_INT_LOCAL_SCOPE))) {
+		return DROP_NO_SERVICE;
+	}
+	/* continue via the slowpath */
+	return CTX_ACT_OK;
+}
```
Bonus: you can use this helper in bpf_sock as well. The call site at bpf_sock and bpf_lxc can return the appropriate error code.

oblazek replied:
true, my bad.. though let me do that in a quick follow-up

aditighag reviewed (approved):
Thanks for the revisions, lgtm! Left one more suggestion for code dedup. 
Edit: Since all tests passed, feel free to address this as a follow-up.

rgo3 reviewed (approved):
LGTM!

rastislavs commented on pkg/bgpv1/manager/reconcilerv2/service.go:
```diff
@@ -367,13 +371,19 @@ func (r *ServiceReconciler) updateServiceAdvertisementsMetadata(p ReconcileParam
 }
 
-func hasLocalBackends(p ReconcileParams, fe *loadbalancer.Frontend) bool {
+// hasLocalBackends loops through Frontend backends and returns:
```
Thanks for the patience and update!

One nit: given the func name (hasLocalBackends), I find it a bit confusing that we now return `true, false` when there are no backends at all. I would suggest returning `false, false` in that case, and maybe using named return values to increase readability, e.g.:

```go
func hasBackends(p ReconcileParams, fe *loadbalancer.Frontend) (hasBackends bool, hasLocalBackends bool) {
```

oblazek replied:
sounds good, let me fix that

rastislavs commented on pkg/bgpv1/manager/reconcilerv2/service.go:
```diff
@@ -367,13 +371,22 @@ func (r *ServiceReconciler) updateServiceAdvertisementsMetadata(p ReconcileParam
+	n := 0
 	for backend := range fe.Backends {
+		n++
+		hasBackends = true
```
last (very small) nit: we don't seem to need the `n`  counter anymore

oblazek replied:
done
